evm/rlp/headers.py

```diff
@@ -196,15 +207,57 @@ def __repr__(self):
 
     @property
     def hash(self):
-        return keccak(rlp.encode(self))
+        int_to_bytes32 = compose(
+            pad32,
+            int_to_big_endian,
+        )
+        header_hash = keccak(
+            b''.join((
+                int_to_bytes32(self.shard_id),
+                int_to_bytes32(self.expected_period_number),
+                self.period_start_prevhash,
+                self.parent_hash,
+                self.transaction_root,
+                pad32(self.coinbase),
+                self.state_root,
+                self.receipt_root,
+                int_to_bytes32(self.number),
+            ))
+        )
+        return header_hash
+
+    @classmethod
+    @to_dict
+    def _deserialize_header_bytes_to_dict(cls, header_bytes):
+        header_sedes = cls.get_sedes()
+        # assume all fields are padded to 32 bytes
+        obj_size = 32
+        assert len(header_bytes) == obj_size * len(header_sedes)
```

Comment:
Assertions are generally bad for runtime code as they can be easily disabled.  This should probably raise a `ValidationError`

Reply:
Have changed it. Thank you!